File: leetcode/go/_015/main.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
	fmt.Println(threeSum([]int{-4, -1, -1, 0, 1, 2}))
}

func threeSum(nums []int) [][]int {
	sort.Ints(nums)

	result := make([][]int, 0)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i-1] == nums[i] {
			continue
		}

		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if sum > 0 {
				r--
			} else if sum < 0 {
				l++
			} else {
				result = append(result, []int{nums[i], nums[l], nums[r]})
				l++
				for nums[l-1] == nums[l] && l < r {
					l++
				}
			}
		}
	}

	return result
}

File: leetcode/python/_015/main.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i-1]: continue
            j, k = i + 1, n - 1
            while j < k:
                c = nums[i] + nums[j] + nums[k]
                if c == 0:
                    res.append((nums[i], nums[j], nums[k]))
                    j += 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                elif c < 0:
                    j += 1
                else:
                    k -= 1
        return res

print(Solution().threeSum([-1, 0, 1, 2, -1, -4]))
print(Solution().threeSum([-4, -1, -1, 0, 1, 2]))